### ISSUE TITLE
Parametrize our regression tests

### ISSUE THREAD

**Issue by jdavies-st** [spacetelescope/jwst#2507]
The vast majority of our regression tests have exactly the same format - a pipeline step is run on an input FITS file, and the output is compared to a truth FITS file.  They are essentially the same test that can be parametrized with different input data, truth data, and pipeline step classes.  In some cases step parameters are set as well.
```
testdata = [
    ("jw10072871_input_rate.fits", "jw10072871_input_assign_wcs.fits", AssignWcsStep, None),
    ("jw12345_nrs1_cal.fits", "jw12345_nrs1_s2d.fits", ResampleSpecStep, dict(save_results=True)),
etc
]
@pytest.mark.parametrize("input_file, truth_file, step_class, step_params", testdata)
```
So instead of having 118 files containing somewhat more tests, we could have for each instrument, one parametrized test that takes care the majority of the pipeline step testing.  One file, one test.

For the pipeline tests where several outputs are compared, there's probably limited gains in parametrizing them.

The goal would be to reduce the lines of code in the tests by an order of magnitude using `@pytest.mark.parametrize()`.  There would in the end be the same number of tests, and they would run independently, just as they do now.  Just less code.

Thoughts?

**Comment by hbushouse** [comment]
I like it! Also then makes it very easy for someone to create a new test. Instead of rewriting from scratch or doing a copy and paste of an existing test module, all they would need to do is add a new "testdata" entry.

**Comment by jdavies-st** [comment]
Fixed by #2840.
